File: mpbeps2/ReadFiles/diag2.f90
!-----------------------------------------------------------------------
! This program reads 2d data written by 2D MPI/OpenMP PIC codes
! written by Viktor K. Decyk, UCLA
      program diag2
      use cmfield2
      use graf2
      use pread2
      implicit none
      integer, parameter :: ms = 6, ns = 7
      integer :: idrun = 0, iudm = 19
      integer :: i, m, n
! mscalars = table of available diagnostic types
      integer, dimension(ms) :: mscalars = 0
! nscalars = table of available diagnostics
      integer, dimension(ns) :: nscalars = 0
      character(len=18), dimension(ms) :: dname = (/'REAL SCALAR FIELDS'&
     &,'REAL VECTOR FIELDS','REAL FLUID DATA   ','VELOCITY DATA     ',  &
     &'TRAJECTORY DATA   ','PHASE SPACE DATA  '/)
      character(len=10) :: cdrun
      character(len=32) :: fname
!
! create string from idrun
      write (*,*) 'enter idrun:'
      read (5,*) idrun
      write (cdrun,'(i10)') idrun
      cdrun = adjustl(cdrun)
      fname = 'diag2.'//cdrun
      call ffopen2(iudm,fname)
!
! determine which scalar diagnostics are available
      call readsdiags2(iudm,nscalars)
      mscalars(1) = sum(nscalars)
! determine which vector diagnostics are available
      nscalars = 0
      call readvdiags2(iudm,nscalars)
      mscalars(2) = sum(nscalars)
! determine which fluid diagnostics are available
      nscalars = 0
      call readfldiags2(iudm,nscalars)
      mscalars(3) = sum(nscalars)
! determine which velocity diagnostics are available
      nscalars = 0
      call readfvdiags2(iudm,nscalars)
      mscalars(4) = sum(nscalars)
! determine which trajectory diagnostics are available
      nscalars = 0
      call readtrdiags2(iudm,nscalars)
      mscalars(5) = sum(nscalars)
! determine which phase space diagnostics are available
      nscalars = 0
      call readpsdiags2(iudm,nscalars)
      mscalars(6) = sum(nscalars)
!
! select diagnostic
      m = sum(mscalars)
      do
         if (m > 0) then
            n = -1
            do while (n < 0)
               do i = 1, ms
                  if (mscalars(i) > 0) then
                     write (*,*) 'enter ', i, 'for ', trim(dname(i))
                  endif
               enddo
               write (*,*) 'enter ', 0, 'for EXIT'
               read (5,*) n
               if (n==0) then
                  call closeff2(iudm)
                  stop
               endif
               if ((n >= 1).and.(n <= ns)) then
                  if (mscalars(n)==0) n = -1
               else
                  n = -1
               endif
               if (n > 0) exit
               write (*,*) 'invalid entry, try again or enter 0 to quit'
            enddo
         else
            write (*,*) 'no diagnostic data found'
            stop
         endif
!
         write (*,*) trim(dname(n)), ' diagnostic selected'
!
         if (n==1) then
            call preadf2(idrun)
         else if (n==2) then
            call pvreadf2(idrun)
         else if (n==3) then
            call pflreadf2(idrun)
         else if (n==4) then
            call pfvreadf2(idrun)
         else if (n==5) then
            call ptreadf2(idrun)   
         else if (n==6) then
            call psreadf2(idrun)
         endif
      enddo
!
      end program

File: mpbeps2/diag2.py
#-----------------------------------------------------------------------
# This program reads 2d data written by 2D MPI/OpenMP PIC codes
# written by Viktor K. Decyk, UCLA
from __future__ import print_function
import sys
import math
import numpy

sys.path.append('./ReadFiles')
from libcmfield2 import *
import preadf2
import pvreadf2
import pflreadf2
import pfvreadf2
import ptreadf2
import psreadf2

if (sys.version_info.major==2):
   input = raw_input

int_type = numpy.int32
double_type = numpy.float64
float_type = numpy.float32
complex_type = numpy.complex64

ms = 6; ns = 7
iudm = 19
nplot = 1
dname = numpy.array(["REAL SCALAR FIELDS","REAL VECTOR FIELDS",
                     "REAL FLUID DATA   ","VELOCITY DATA     ",
                     "TRAJECTORY DATA   ","PHASE SPACE DATA  "],
                     dtype=str)

# create string from idrun
cdrun = "Unknown"
while (cdrun.isdigit() == False):
   cdrun = input("enter integer idrun: ")
idrun = int(cdrun)
fname = "diag2." + cdrun
cmfield2.ffopen2(iudm,fname)

# mscalars = table of available diagnostic types
mscalars = numpy.zeros((ms),int_type,'F')
# nscalars = table of available diagnostics
nscalars = numpy.zeros((ns),int_type,'F')

# determine which scalar diagnostics are available
cmfield2.readsdiags2(iudm,nscalars)
mscalars[0] = numpy.sum(nscalars)
# determine which vector diagnostics are available
nscalars[:] = 0
cmfield2.readvdiags2(iudm,nscalars)
mscalars[1] = numpy.sum(nscalars)
# determine which fluid diagnostics are available
nscalars[:] = 0
cmfield2.readfldiags2(iudm,nscalars)
mscalars[2] = numpy.sum(nscalars)
# determine which velocity diagnostics are available
nscalars[:] = 0
cmfield2.readfvdiags2(iudm,nscalars)
mscalars[3] = numpy.sum(nscalars)
# determine which trajectory diagnostics are available
nscalars[:] = 0
cmfield2.readtrdiags2(iudm,nscalars)
mscalars[4] = numpy.sum(nscalars)
# determine which phase space diagnostics are available
nscalars[:] = 0
cmfield2.readpsdiags2(iudm,nscalars)
mscalars[5] = numpy.sum(nscalars)

# select diagnostic
m = numpy.sum(mscalars)
while True:
   if (m > 0):
      n = -1
      while True:
         if (n < 0):
            for i in range(0,ms):
               if (mscalars[i] > 0):
                  print ("enter ", i+1," for", numpy.str.rstrip(dname[i]))
            print ("enter ", 0, " for EXIT")
            c = input("")
            if (c.isdigit()):
               n = int(c)
            if (n==0):
               exit(1)
            if ((n >= 1) and (n <= ms)):
               if (mscalars[n-1]==0):
                  n = -1
            else:
               n = -1
            if (n > 0):
               break
            print ("invalid entry, try again or enter 0 to quit")
   else:
      print ("no diagnostic data found")
      exit(1)

   print (numpy.str.rstrip(dname[n-1]), " diagnostic type selected")

   if (n==1):
      preadf2.main(idrun)
   elif (n==2):
      pvreadf2.main(idrun)
   elif (n==3):
      pflreadf2.main(idrun)
   elif (n==4):
      pfvreadf2.main(idrun)
   elif (n==5):
      ptreadf2.main(idrun)
   elif (n==6):
      psreadf2.main(idrun)

cmfield2.closeff2(iudm)

